src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,134 @@
+"""
+Perform Chavatal-Gomory strengthening of inequalities
+
+We apply the Chvatal-Goromory procedure to constraints that
+satisfy the following criteria:
+1. All variables with nonzero coefficient must be integral
+2. The constraints are not ranged
+3. If the constraint is of the form ax ⩾ b, then every
+   positive coefficient variable must have finite lowerbound and
+   negative coefficient variables must have finite upperbound.
+
+   If the constraint is of the from ax ⩽ b, then every
+   positive coefficient variable must have finite upperbound and
+   negative coefficient variables must have finite lowerbound.
+
+For example, let ax ⩾ b be a constraint where aᵢ ⩾ 0,
+and every non-zeoro coefficient in a corresponds to an
+integer variable. If there exists an s > 0 such that
+⌈aⱼ s⌉ b / ⌈b s⌉ ⩽ aⱼ
+and there is k such that
+⌈aₖ s⌉ b / ⌈b s⌉ < aₖ (*)
+then we will update ax ⩾ b into ⌈a s⌉ x ⩾ ⌈b s⌉.
+
+We find such a number s heuristically, i.e we try different
+values from the set
+{1, t/a_max, t/a_min, (2t-1)/(2a_min)| t = 1,..., r}
+where a_max = sup(|aᵢ|), a_min = inf(|aᵢ|) and r is resolution which
+is set to be 5 by default.
+
+When performing CG strengthening inequality, we round up positive
+coefficients and round down negative ones, so this procedure
+will not reduce any coefficient to 0.
+"""
+
+function cg_strengthening!(ps::PresolveData{T}, resolution::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+    for i in 1:ps.pb0.ncon
+        ps.rowflag[i] || continue
+        nzind = ps.pb0.arows[i].nzind
+        nzval = ps.pb0.arows[i].nzval
+
+        if isfinite(ps.urow[i]) && isfinite(ps.lrow[i]) ||
```

Comment:
It is slightly preferred to define booleans for these two statements as they are used in the checks for different branches. If this were a hot loop (it's not) this could improve performance, but it also helps ensure that the logic doesn't get out of sync later on.

E.g.:
```jl
finite_lb = isfinite(ps.lrow[i])
finite_ub = isfinite(ps.urow[i])
if finite_lb && finite_ub
    ...
elseif finite_lb
    ...
elseif finite_ub
    ...
else
    error("How did we get an unbounded constraint in here???")
end
```

Reply:
I think this actually should be ps.pb0.ncon instead of ps.nrow. Because, we only mark which row is not active without deleting it from ps.pb0.arows, ps.nrow maybe out of sync with ps.pb0.arows.

In my test4, I deactivate 2 first constraints, and there will be error if I used ps.nrow. Because, in this case ps.nrow is 1, however, we need to access the third row in ps.pb0.arow to look for CG inequality.